src/nuklear_font.c

```diff
@@ -376,1689 +17,33 @@ nk_rp_pack_rects(struct nk_rp_context *context, struct nk_rp_rect *rects, int nu
-}
-NK_INTERN int
-nk_tt__GetGlyfOffset(const struct nk_tt_fontinfo *info, int glyph_index)
-{
-    int g1,g2;
-    if (glyph_index >= info->numGlyphs) return -1; /* glyph index out of range */
-    if (info->indexToLocFormat >= 2)    return -1; /* unknown index->glyph map format */
-
-    if (info->indexToLocFormat == 0) {
-        g1 = info->glyf + nk_ttUSHORT(info->data + info->loca + glyph_index * 2) * 2;
-        g2 = info->glyf + nk_ttUSHORT(info->data + info->loca + glyph_index * 2 + 2) * 2;
-    } else {
-        g1 = info->glyf + (int)nk_ttULONG (info->data + info->loca + glyph_index * 4);
-        g2 = info->glyf + (int)nk_ttULONG (info->data + info->loca + glyph_index * 4 + 4);
-    }
-    return g1==g2 ? -1 : g1; /* if length is 0, return -1 */
-}
-NK_INTERN int
-nk_tt_GetGlyphBox(const struct nk_tt_fontinfo *info, int glyph_index,
-    int *x0, int *y0, int *x1, int *y1)
-{
-    int g = nk_tt__GetGlyfOffset(info, glyph_index);
-    if (g < 0) return 0;
-
-    if (x0) *x0 = nk_ttSHORT(info->data + g + 2);
-    if (y0) *y0 = nk_ttSHORT(info->data + g + 4);
-    if (x1) *x1 = nk_ttSHORT(info->data + g + 6);
-    if (y1) *y1 = nk_ttSHORT(info->data + g + 8);
-    return 1;
-}
-NK_INTERN int
-nk_tt__close_shape(struct nk_tt_vertex *vertices, int num_vertices, int was_off,
-    int start_off, nk_int sx, nk_int sy, nk_int scx, nk_int scy, nk_int cx, nk_int cy)
-{
-   if (start_off) {
-      if (was_off)
-         nk_tt_setvertex(&vertices[num_vertices++], NK_TT_vcurve, (cx+scx)>>1, (cy+scy)>>1, cx,cy);
-      nk_tt_setvertex(&vertices[num_vertices++], NK_TT_vcurve, sx,sy,scx,scy);
-   } else {
-      if (was_off)
-         nk_tt_setvertex(&vertices[num_vertices++], NK_TT_vcurve,sx,sy,cx,cy);
-      else
-         nk_tt_setvertex(&vertices[num_vertices++], NK_TT_vline,sx,sy,0,0);
-   }
-   return num_vertices;
-}
-NK_INTERN int
-nk_tt_GetGlyphShape(const struct nk_tt_fontinfo *info, struct nk_allocator *alloc,
-    int glyph_index, struct nk_tt_vertex **pvertices)
-{
-    nk_short numberOfContours;
-    const nk_byte *endPtsOfContours;
-    const nk_byte *data = info->data;
-    struct nk_tt_vertex *vertices=0;
-    int num_vertices=0;
-    int g = nk_tt__GetGlyfOffset(info, glyph_index);
-    *pvertices = 0;
-
-    if (g < 0) return 0;
-    numberOfContours = nk_ttSHORT(data + g);
-    if (numberOfContours > 0) {
-        nk_byte flags=0,flagcount;
-        nk_int ins, i,j=0,m,n, next_move, was_off=0, off, start_off=0;
-        nk_int x,y,cx,cy,sx,sy, scx,scy;
-        const nk_byte *points;
-        endPtsOfContours = (data + g + 10);
-        ins = nk_ttUSHORT(data + g + 10 + numberOfContours * 2);
-        points = data + g + 10 + numberOfContours * 2 + 2 + ins;
-
-        n = 1+nk_ttUSHORT(endPtsOfContours + numberOfContours*2-2);
-        m = n + 2*numberOfContours;  /* a loose bound on how many vertices we might need */
-        vertices = (struct nk_tt_vertex *)alloc->alloc(alloc->userdata, 0, (nk_size)m * sizeof(vertices[0]));
```

Comment:
The adapted `stb_truetype` library in Nuklear used to use the `nk_allocator` that `nk_font_baker` (which comes from `nk_font_atlas`) uses for all its memory allocations.
Removing these customizations, and insisting on not introducing them, has forced the use of separate methods for memory management in `stb_truetype`:
```C
#ifndef STBTT_malloc
static nk_handle fictional_handle = {0};

#define STBTT_malloc(x,u)  nk_malloc( fictional_handle, 0, x )
#define STBTT_free(x,u)    nk_mfree( fictional_handle , x)
#endif
```

`nk_malloc` and `nk_mfree` only exist when `NK_INCLUDE_DEFAULT_ALLOCATOR` is defined.

So the negative effect of this change is that the library no longer builds when `NK_INCLUDE_DEFAULT_ALLOCATOR` is not defined.

This presents a problem for libraries that wrap Nuklear and provide their own memory management, eg many bindings for other languages.


As an example, I am working on updating a binding from a Nuklear version before this PR to a Nuklear version after this PR.
It is preferred to not let Nuklear pull in stdlib for memory management, since the language in question (Rust, but the point of this comment is language-agnostic) can provide its own, and it makes no sense to have two separate ways of managing memory.
This was previously possible by creating a `nk_allocator` that points to custom alloc/free implementations that delegate to the language's implementations, and initializing a `nk_font_atlas` with this allocator; this allocator would be used for all font-related memory management in Nuklear, including in `stb_truetype`.

After this PR, `stb_truetype` no longer uses the `nk_allocator` of `nk_font_atlas`.
This means it must be provided with a separate implementation in the binding code.
This is not technically difficult (just implement a malloc and a free, and alias `STBTT_malloc` and `STBTT_free` to them), but is certainly cumbersome for the users of this library that actually prefer finer control over memory management (which, given the library's lightweight nature, is more than just those using bindings to other languages).

Another subtle difference is that, once `STBTT_malloc` and `STBTT_free` are dealt with as above, it is no longer possible to use two `nk_font_atlas` with two different allocators at the same time in the same manner as before this change, since `stb_truetype`'s memory management cannot switch between different allocators at run time (without manually implementing that, and introducing locking for multithreaded accesses if necessary, etc).
Using multiple `nk_font_atlas` in the same application simultaneously is not something that I have personally needed so far, and can't immediately think of a reason where this would be needed, but it is certainly worth noting.

Reply:
@asoft thanks for coming here and describing in detail what this change caused. What would you suggest we implement/change?

Btw. we're open to PRs (especially those increasing interoperability without much cost), so feel free to make one and we might discuss the details there.

Reply:
If allowing for finer control over mememory management is a goal of Nuklear, I think the current situation around memory management in stb_tt should be treated as a bug.
I'd like this bug to be fixed.

Options I see, in order of perceived implementation difficulty:
1. Add configuration to setup stb_tt library-wide to use a given allocator (via an alternative implementation of `STBTT_malloc` and `STBTT_free`).
1. Propagate `nk_font_atlas`'s allocator to stb_tt by _modifying Nuklear's copy of stb_tt_, adding the allocator parameter in all methods where `STBTT_malloc`/`STBTT_free` are used
1. Propagate `nk_font_atlas`'s allocator to stb_tt or `STBTT_malloc`/`STBTT_free` via some other means, without modifying the committed copy of stb_tt

Pros/cons that I see:
1. pros: simple to implement; simple to use\
cons: introduces yet another place for memory management; has all drawbacks mentioned in initial comment.
1. pros: simple to implement; simple to use; does not change amount of various memory management entry points \
cons: complicates stb_tt version updates
1. pros: simple to use; does not change amount of various memory management entry points \
cons: likely complicated implementation, cannot be implemented immediately, need to research options; 

I favor option 3, because it does not have the cons of options 1 and 2.
I'm playing locally with some options, don't have a good solution yet.

Reply:
Hm, how about reversing the dependency and propagate stb_tt's `STBTT_malloc/STBTT_free` to nk_font_atlas or something along these lines?

Reply:
I found a simple option 3 solution: #186.